pkg/server/registry/images/tag.go

```diff
@@ -58,37 +59,54 @@ func (t *TagStrategy) ImageTag(ctx context.Context, namespace, imageName string,
 	if err != nil {
 		return nil, err
 	}
+	res, err := normalizeTags(image.Tags)
+	if err != nil {
+		return nil, err
+	}
+	set := sets.NewString(res...)
 
-	duplicateTag := make(map[string]bool)
-	for _, tag := range tags {
-		imageParsedTag, err := name.NewTag(tag, name.WithDefaultRegistry(""))
-		if err != nil {
-			return nil, err
-		}
-		duplicateTag[imageParsedTag.Name()] = true
+	imageParsedTag, err := name.NewTag(tagToAdd, name.WithDefaultRegistry(""))
+	if err != nil {
+		return nil, err
 	}
+	set.Insert(imageParsedTag.Name())
 
+	hasChanged := false
 	for _, img := range imageList.Items {
-		if img.Name == image.Name {
-			continue
+		res, err = normalizeTags(img.Tags)
+		if err != nil {
+			return nil, err
 		}
-		for i, tag := range img.Tags {
-			imageParsedTag, err := name.NewTag(tag, name.WithDefaultRegistry(""))
-			if err == nil {
-				tag = imageParsedTag.Name()
-			} else {
-				logrus.Errorf("invalid tag [%s] found, ignoring: %v", tag, err)
+		for i, tag := range res {
+			if img.Name == image.Name {
```

Comment:
this check should go outside this inner loop. if `img.Name == image.Name`, we should never have to enter this inner for loop. as the logic currently stands, you would still go through the loop the maximum number of times.